perllib/FixMyStreet/App/Controller/Develop.pm

```diff
@@ -0,0 +1,136 @@
+        'problem-confirm' => 1, 'problem-confirm-not-sending' => 1,
+        'problem-moderated' => 1, 'questionnaire' => 1, 'submit' => 1);
+
+    my $update = $c->model('DB::Comment')->first;
+    my $problem = $c->model('DB::Problem')->first;
+
+    $c->stash->{templates} = [];
+    foreach (sort keys %templates) {
+        my $url = $_;
+        $url .= "?problem=" . $problem->id if $problem && $with_problem{$_};
+        $url .= "?update=" . $update->id if $update && $with_update{$_};
+        push @{$c->stash->{templates}}, { name => $_, url => $url };
+    }
+}
+
+=item email_previewer
+
+Previews an HTML email template. A problem or update ID can be provided as a
+query parameter, and other data is taken from the database.
+
+=back
+
+=cut
+
+sub email_previewer : Path('/_dev/email') : Args(1) {
+    my ( $self, $c, $template ) = @_;
+
+    my $vars = {};
+    if (my $id = $c->get_param('update')) {
+        $vars->{update} = $c->model('DB::Comment')->find($id);
+        $vars->{problem} = $vars->{report} = $vars->{update}->problem;
+    } elsif ($id = $c->get_param('problem')) {
+        $vars->{problem} = $vars->{report} = $c->model('DB::Problem')->find($id);
+    }
+
+    # Special case needed variables
+    if ($template =~ /^alert-problem/) {
+        $vars->{area_name} = 'Area Name';
+        $vars->{ward_name} = 'Ward Name';
+        $vars->{data} = [ $c->model('DB::Problem')->search({}, { rows => 5 })->all ];
+    } elsif ($template eq 'alert-update') {
+        $vars->{data} = [];
+        my $q = $c->model('DB::Comment')->search({}, { rows => 5 });
+        while (my $u = $q->next) {
+            my $fn = sub {
+                return FixMyStreet::App::Model::PhotoSet->new({
+                    db_data => $u->photo,
+                })->get_image_data( num => 0, size => 'fp' );
+            };
+            push @{$vars->{data}}, {
+                item_photo => $u->photo, get_first_image_fp => $fn, item_text => $u->text,
+                item_name => $u->name, item_anonymous => $u->anonymous, confirmed => $u->confirmed };
+        }
+    } elsif ($template eq 'questionnaire') {
+        $vars->{created} = 'N weeks';
+    }
+
+    my $email = $c->construct_email("$template.txt", $vars);
+
+    # Look through the Email::MIME email for the text/html part, and any inline
+    # images. Turn the images into data: URIs.
+    my $html = '';
+    my %images;
+    $email->walk_parts(sub {
+        my ($part) = @_;
+        return if $part->subparts;
+        if ($part->content_type =~ m[^image/]i) {
+            (my $cid = $part->header('Content-ID')) =~ s/[<>]//g;
+            (my $ct = $part->content_type) =~ s/;.*//;
+            $images{$cid} = "$ct;base64," . $part->body_raw;
+        } elsif ($part->content_type =~ m[text/html]i) {
+            $html = $part->body_str;
+        }
+    });
+
+    foreach (keys %images) {
+        $html =~ s/cid:([^"]*)/data:$images{$1}/g;
+    }
+
+    $c->response->body($html);
```

Comment:
Something is injecting jQuery and the plDebug toolbar into the response that my browser is receiving from the server. I guess it _shouldn’t_ matter, because neither of these things should affect the layout of the emails, but I just wondered whether it would be more _theoretically pure_ if the HTML we received in this preview was _literally_ just the HTML of the HTML part of the email, and nothing else?

Reply:
You can run `FIXMYSTREET_APP_DEBUG=0 script/server` to run the development server with debug switched off (including the toolbar). As you say, it shouldn't affect it (and might be useful to see what queries are being made just in the render, perhap). The toolbar is looking for `</body>` and attaching itself, as it operates 'above' the main code I don't think there's a way for the code to disable it for a particular route (there probably is somehow, but can't see one obviously).

Reply:
Yes, you’re right, there might be tools in the debug toolbar that you want to use to interrogate the email. (Even if not right now, maybe in the future!). Good tip about disabling `FIXMYSTREET_APP_DEBUG` at runtime. :+1: